### PR TITLE
fix(tooltip): render tooltip on portal to avoid hidden overflows

### LINKED ISSUE #375: Enhance tooltip placement
**Is your feature request related to a problem? Please describe.**
Enhance tooltip placement to avoid janky placement especially on smaller charts.

Screen Recording 2019-09-12 at 08 59 PM

**Describe the solution you'd like**
Have a bounding box tooltip that jumps only when needed (i.e. beyond the chart grids, beyond chart or beyond the browser window)

**Kibana Cross Issues**
Add any [Kibana](https://github.com/elastic/kibana) related issues/feature requests here.

**Checklist**

- [ ] this request is checked against already exist requests
- [ ] every related Kibana issue is listed under `Kibana Cross Issues` list
- [ ] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### PR DESCRIPTION
**Pull request by markov00**
## Summary

The current tooltip was rendered inside the chart container. If the chart is rendered inside a DOM
element with an hidden overflow, then the tooltip will be clipped. I moved the tooltip element to
render on the document body through a React portal.

The tooltip is generally always displayed on the right of the crosshair/crosshair band. If it overflow the chart width, height, depending on the rotation, it will stop moving or will jump on the left side of the band 

close #375

Oct-15-2019 14-33-56

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Proper documentation or storybook story was added for features that require explanation or tutorials
- [x] Unit tests were updated or added to match the most common scenarios
- [x] Each commit follows the [convention](https://github.com/elastic/elastic-charts/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by markov00** [comment]
> This isn't blocking but I was curious (and ignorant :D) about the changes in removing lines 71-73 in common.ts - I'm assuming `await this.loadChartFromUrl(url)` and `await this.waitForChartRendered()` is just an updated version of `CommonPage.parseUrl(url)`?

Yes, I moved the `parseUrl` out from `expectChartAtUrlToMatchScreenshot ` so that I can reuse it easily on other places. The same thing for selector that waits for the chart to be present on the DOM

File: src/components/tooltips.tsx [hunk 1]
```diff
@@ -1,15 +1,59 @@
...
   static displayName = 'Tooltips';
+  portalNode: HTMLDivElement | null = null;
+  tooltipRef: React.RefObject<HTMLDivElement>;
+
+  constructor(props: TooltipProps) {
+    super(props);
+    this.tooltipRef = React.createRef();
+  }
+  componentDidMount() {
+    const container = document.body.getElementsByClassName('echTooltipContainer')[0];
```
Comment (nickofthyme) [inline src/components/tooltips.tsx]: this should be by id no? maximize specificity

File: src/components/tooltips.tsx [hunk 2]
```diff
@@ -1,15 +1,59 @@
...
+    if (container) {
+      this.portalNode = container as HTMLDivElement;
+    } else {
+      this.portalNode = document.createElement('div');
+      this.portalNode.className = 'echTooltipContainer';
```
Comment (nickofthyme) [inline src/components/tooltips.tsx]: id?

Comment (markov00) [reply src/components/tooltips.tsx]: fix on https://github.com/elastic/elastic-charts/pull/418/commits/c3fcec92ba2109376526ec0562d22a572b034cf6

File: src/components/tooltips.tsx (hunk 1 above)
Comment (markov00) [reply src/components/tooltips.tsx]: fix on https://github.com/elastic/elastic-charts/pull/418/commits/c3fcec92ba2109376526ec0562d22a572b034cf6

In that commit I'm also assuring that the portalNode is recreated if another chart removed it after unmounting that chart

File: src/chart_types/xy_chart/crosshair/crosshair_utils.ts [hunk 3]
```diff
@@ -8,6 +8,17 @@ export interface SnappedPosition {
   position: number;
   band: number;
 }
+export interface TooltipPosition {
+  isHorizontalRotated: boolean;
+  vPosition: {
+    bandHeight: number;
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/crosshair/crosshair_utils.ts]: added jsdocs in https://github.com/elastic/elastic-charts/pull/418/commits/25789aabd369f9295d8892c32c02d88cdabb9b88

File: src/chart_types/xy_chart/crosshair/crosshair_utils.ts [hunk 4]
```diff
@@ -8,6 +8,17 @@ export interface SnappedPosition {
   position: number;
   band: number;
 }
+export interface TooltipPosition {
+  isHorizontalRotated: boolean;
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/crosshair/crosshair_utils.ts]: renamed in https://github.com/elastic/elastic-charts/pull/418/commits/7005c7a263294355812bd08f2f9603550371ab23

File: integration/page_objects/common.ts [hunk 5]
```diff
@@ -81,6 +111,14 @@ class CommonPage {
       throw new Error(error);
     }
   }
+  async loadChartFromURL(url: string) {
+    const cleanUrl = CommonPage.parseUrl(url);
+    await page.goto(cleanUrl);
+  }
+
+  async waitForChartRendered(selector = '.echChart[data-ech-render-complete=true]', timeout = 10000) {
```
[the remark this answers is not in the document]
Comment (markov00) [reply integration/page_objects/common.ts]: renamed in https://github.com/elastic/elastic-charts/pull/418/commits/182ef4431b71edc1aa3f3ff756e9a281b293e8cc

File: src/chart_types/xy_chart/crosshair/crosshair_utils.ts [hunk 6]
```diff
@@ -9,13 +9,18 @@ export interface SnappedPosition {
...
+    // the top position of the tooltip relative to the parent
     bandTop: number;
+    // the height of the crosshair band if any
+    bandHeight: number;
   };
   hPosition: {
+    // the left position of the tooltip relative to the parent
     bandLeft: number;
+    // the width of the crosshair band if any
     bandWidth: number;
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/crosshair/crosshair_utils.ts]: soved in faf1552

File: integration/page_objects/common.ts [hunk 7]
```diff
@@ -52,11 +56,13 @@ class CommonPage {
...
-   * Capture screenshot or chart element only
-   */
-  async getChartScreenshot() {
-    return this.screenshotDOMElement('.echChart[data-ech-render-complete=true]');
+  async moveMouseRelativeToDOMElement(
+    mousePosition: { x: number; y: number },
+    selector = '.echChart[data-ech-render-complete=true]',
+  ) {
+    const chartContainer = await this.getBoundingClientRect(selector);
+    console.log({ chartContainer });
```
Comment (nickofthyme) [inline integration/page_objects/common.ts]: 😲

File: src/chart_types/xy_chart/annotations/annotation_tooltip.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  tooltipAnchor: { top: number; left: number },
+  padding = 10,
+): {
+  left: string | null;
+  top: string | null;
+} {
+  let left = 0;
+
+  const annotationXOffset = window.pageXOffset + container.left + chartDimensions.left + tooltipAnchor.left;
+  console.log({
```
Comment (nickofthyme) [inline src/chart_types/xy_chart/annotations/annotation_tooltip.ts]: 😲

Comment (nickofthyme) [reply src/chart_types/xy_chart/annotations/annotation_tooltip.ts]: https://github.com/elastic/elastic-charts/pull/444

**Pull request merged**